clicommand/pipeline_upload.go

```diff
@@ -227,35 +233,46 @@ var PipelineUploadCommand = cli.Command{
 			l.Fatal("Pipeline parsing of \"%s\" failed (%s)", src, err)
 		}
 
-		// In dry-run mode we just output the generated pipeline to stdout
-		if cfg.DryRun {
-			enc := json.NewEncoder(os.Stdout)
-			enc.SetIndent("", "  ")
-
-			// Dump json indented to stdout. All logging happens to stderr
-			// this can be used with other tools to get interpolated json
-			if err := enc.Encode(result); err != nil {
-				l.Fatal("%#v", err)
-			}
-
-			return
-		}
-
 		if len(cfg.RedactedVars) > 0 {
 			needles := redaction.GetKeyValuesToRedact(shell.StderrLogger, cfg.RedactedVars, env.FromSlice(os.Environ()).ToMap())
 			serialisedPipeline, err := result.MarshalJSON()
 
 			if err != nil {
-				l.Fatal("Couldn’t scan the %q pipeline for redacted variables. This parsed pipeline could not be serialized, ensure the pipeline YAML is valid, or ignore interpolated secrets for this upload by passing --redacted-vars=''. (%s)", src, err)
+				l.Fatal("Couldn't scan the %q pipeline for redacted variables. This parsed pipeline could not be serialized, ensure the pipeline YAML is valid, or ignore interpolated secrets for this upload by passing --redacted-vars=''. (%s)", src, err)
 			}
 
 			stringifiedserialisedPipeline := string(serialisedPipeline)
 
+			secretsFound := make([]string, 0, len(needles))
 			for needleKey, needle := range needles {
 				if strings.Contains(stringifiedserialisedPipeline, needle) {
-					l.Fatal("Couldn’t upload the %q pipeline. Refusing to upload pipeline containing the value interpolated from the %q environment variable. Ensure your pipeline does not include secret values or interpolated secret values.", src, needleKey)
+					secretsFound = append(secretsFound, needleKey)
 				}
 			}
+
+			if len(secretsFound) > 0 {
+				if cfg.RejectSecrets {
+					l.Fatal("Refusing to upload pipeline %q, as it contained values interpolated from the following secret environment variables: %v. Ensure your pipeline does not include secret values or interpolated secret values.", src, secretsFound)
+				} else {
+					l.Warn("The pipeline %q contains values interpolated from the following secret environment variables: %v.", src, secretsFound)
+					l.Warn("I'm still going to upload the pipeline, but if you'd like to to prevent this from happening, you can use the `--reject-secrets` cli flag, or the `BUILDKITE_PIPELINE_REJECT_SECRETS` environment variable, which will make the `buildkite-agent pipeline upload` command fail if it finds secrets in the pipeline.")
```

Comment:
how do we feel about first-person error messages? i personally like them (QED, i wrote this), but if that's not the ✨buildkite agent vibe✨ then i'm happy to edit this

Reply:
Hmm, doesn't seem quite right to my eye. Maybe more a "This pipeline will still be uploaded"

Reply:
Yea, I'd prefer we didn't use first person here :)

Reply:
cool! no worries, i'll use `this pipeline will still be uploaded`.

Reply:
🤔 i might update the error message when we refuse to update as well, just so the wording is a little more consistent.